sdk/src/main/java/org/coralibre/android/sdk/internal/matching/ExposureUtils.java

```diff
@@ -0,0 +1,295 @@
+
+import org.coralibre.android.sdk.fakegms.nearby.exposurenotification.ExposureConfiguration;
+import org.coralibre.android.sdk.internal.EnFrameworkConstants;
+import org.coralibre.android.sdk.internal.TracingService;
+import org.coralibre.android.sdk.internal.datatypes.DiagnosisKey;
+import org.coralibre.android.sdk.internal.device_info.DeviceInfo;
+import org.coralibre.android.sdk.internal.matching.intermediateDatatypes.Exposure;
+import org.coralibre.android.sdk.internal.matching.intermediateDatatypes.ExposureRecord;
+import org.coralibre.android.sdk.internal.matching.intermediateDatatypes.Match;
+import org.coralibre.android.sdk.internal.matching.intermediateDatatypes.Period;
+import org.coralibre.android.sdk.internal.matching.intermediateDatatypes.TimeAndAttenuation;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Date;
+import java.util.LinkedList;
+import java.util.List;
+
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class ExposureUtils {
+
+
+    /**
+     * Collects consecutive matches from all matches for a single rpik and creates Exposure
+     * objects from these blocks of consecutive matches.
+     * @param matches an (unordered) list of matches for the same rpik
+     * @return a list of Exposure objects for that rpik
+     */
+    protected static List<Exposure> exposuresFromMatches(
+        final LinkedList<Match> matches,
+        final ExposureConfiguration exposureConfiguration,
+        final DiagnosisKey diagnosisKey,
+        final DeviceInfo ownDeviceInfo
+    ) {
+        // First sort the matches by time to ease assembling of the exposures. The sorted
+        // list will be in chronological order, i.e. the most recent match will be the last
+        // list entry:
+        final List<Match> matchesSorted = (LinkedList<Match>) matches.clone();
+        Collections.sort(matches, new Comparator<Match>() {
+            @Override
+            public int compare(Match o1, Match o2) {
+                if (o1.captureTimestampMillis > o2.captureTimestampMillis) {
+                    return 1;
+                }
+                if (o1.captureTimestampMillis < o2.captureTimestampMillis) {
+                    return -1;
+                }
+                return 0;
+            }
+        });
+
+        // Now assemble Exposure objects from consecutive matches that have only a small difference
+        // in the capture time:
+
+        // ... for that, we first split the match list to get one list per exposure:
+        LinkedList<LinkedList<Match>> matchesPerExposure = new LinkedList<>();
+        final long lastMatchTimestampMillis = -1;
+        for (Match match : matchesSorted) {
+            if (lastMatchTimestampMillis == -1
+                || (match.captureTimestampMillis - lastMatchTimestampMillis) * 1000
+                > EnFrameworkConstants.MAX_EXPOSURE_INTERPOLATION_DURATION_SECONDS) {
```

Comment:
This looks strange to me: a microseconds variable (it was millis, then by multiplying by 1000 it became micros) is compared to a seconds one.

Reply:
Fixed